tools/GulpCoverageUtils.ts

```diff
@@ -0,0 +1,129 @@
+                    return GulpCoverageUtils.copyGeneratedCoverageJson(moduleName, modulesRoot, coverageOutputDir);
+                }));
+            })
+            .then(function(): Q.Promise<any> {
+                var coverageResults: string[] = [coverageOutputDir];
+                if (otherCoverageDir) {
+                    coverageResults.push(otherCoverageDir);
+                }
+                return GulpCoverageUtils.mergeAndGenerateReports(coverageResults, coverageOutputDir, modulesRoot);
+            });
+    }
+
+    private static mergeAndGenerateReports(coverageResultsPaths: string[], coverageOutputDir: string, modulesRoot: string): Q.Promise<any> {
+        return GulpCoverageUtils.collectCoverageJsons(coverageResultsPaths)
+            .then(function(coverageJsonsList: string[]): any {
+                return GulpCoverageUtils.normalizeCoveragePaths(coverageJsonsList, coverageOutputDir, modulesRoot);
+            })
+            .then(function(coverageJsonsList: string[]): void {
+                // collect all generated coverage.json and use them to remap to typescript sources
+                var loadCoverage: any = require("remap-istanbul/lib/loadCoverage");
+                var remap: any = require("remap-istanbul/lib/remap");
+                var writeReport: any = require("remap-istanbul/lib/writeReport");
+
+                var coverage = loadCoverage(coverageJsonsList);
+                var collector = remap(coverage, { basePath: "./" });
+                return writeReport(collector, GulpCoverageUtils.CoverageReportType, coverageOutputDir);
+            });
+    }
+
+    private static collectCoverageJsons(coverageResultsPaths: string[]): Q.Promise<string[]> {
+        // read all the directories and prepare a flat list of coverageXXX.json we need to remap
+        return Q.all<string[]>(coverageResultsPaths.map(coverageResultsPath => {
+            coverageResultsPath = path.resolve(coverageResultsPath);
+            return Q.denodeify(fs.readdir)(coverageResultsPath)
+                .then(function(files: string[]): string[] {
+                    return files.filter(file => {
+                        return file.toLowerCase().indexOf("coverage-") === 0 && path.extname(file) === ".json";
+                    }).map(file => path.join(coverageResultsPath, file));
+                });
+        })).spread(function(...args: string[][]): string[] {
+            return args.reduce((coverageFilesList: string[], coverageFiles: string[]) => {
+                return coverageFilesList.concat(coverageFiles);
+            });
+
+        });
+    }
+
+    private static normalizeCoveragePaths(coverageJsonsList: string[], coverageOutputDir: string, modulesRoot: string): Q.Promise<string[]> {
+        var index: number = 0;
+        // put all extra coverage files in a random location under coverage
+        var randomDir: string = Math.floor(Math.random() * 10000) + 1000 + "";
```

Comment:
removed the random dir and replaced it with "secondary". something we delete post coverage